android/libraries/rib-badoo/src/main/java/com/badoo/common/rib/routing/backstack/BackStackManager.kt

```diff
@@ -0,0 +1,186 @@
+        override fun invoke(state: State<C>, action: Action<C>): Observable<out Effect<C>> =
+            when (action) {
+                is ReattachRibsOfLastEntry -> Observable.defer {
+                    return@defer just(
+                        connector.enter(
+                            state.backStack.last()
+                        )
+                    ).flatMap { empty<Effect<C>>() }
+                }
+
+                is Execute -> {
+                    when (val wish = action.wish) {
+                        is Replace -> when {
+                            wish.configuration != state.backStack.last() ->
+                                connector.switchToNew(
+                                    state.backStack,
+                                    wish.configuration,
+                                    detachStrategy = DESTROY
+                                ).flatMap {
+                                    just(Effect.Replace(it.second))
+                                }
+
+                            else -> empty()
+                        }
+
+                        is Push -> when {
+                            wish.configuration != state.backStack.last() ->
+                                connector.switchToNew(
+                                    state.backStack,
+                                    wish.configuration,
+                                    detachStrategy = DETACH_VIEW
+                                ).flatMap {
+                                    just(Effect.Push(it.first!!, it.second))
+                                }
+
+                            else -> empty()
+                        }
+
+                        is NewRoot ->
+                            connector.switchToNew(
+                                state.backStack,
+                                wish.configuration,
+                                detachStrategy = DESTROY
+                            ).flatMap {
+                                just(Effect.NewRoot(it.second))
+                            }
+
+                        is Pop -> when {
+                            state.canPop ->
+                                connector.switchToPrevious(state.backStack, detachStrategy = DESTROY).flatMap {
+                                    just(Effect.Pop(it.second))
+                                }
+                            else -> empty()
+                        }
+
+                        is ShrinkToBundles -> connector.shrinkToBundles(state.backStack).flatMap {
+                            just(Effect.UpdateBackStack(it))
+                        }
+
+                        is TearDown -> Observable.defer {
+                            state.backStack.lastOrNull()?.routingAction?.onLeave()
+                            return@defer empty<Effect<C>>()
+                        }
+                    }
+                }
+            }
+    }
+
+    class ReducerImpl<C : Parcelable> : Reducer<State<C>, Effect<C>> {
```

Comment:
Can be object

Reply:
Actually it can't be since it is generic :(